[CShuwen] iP

CShuwen opened the pull request:
## DukePro

### DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- FAST SUPER FAST to use

### All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

### Features:
 - [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

Commit:
Add Gradle support

CShuwen committed:
Level1, greet, Echo and Exit

CShuwen committed:
Level2, add and list

CShuwen committed:
Level 3 Mark as Done

CShuwen committed:
Level 4 ToDos, Events, Deadlines

CShuwen committed:
A-TextUiTesting

CShuwen committed:
A-TextUiTesting

CShuwen committed:
A-TextUiTesting

CShuwen committed:
Level5 Handle Errors

CShuwen committed:
Level6 Delete

CShuwen committed:
Level 7 Create class database, allow tasks to be stored locally.

CShuwen committed:
correct errors in printing responses

CShuwen committed:
level 7- update delete, mark and unmark after storing tasks data locally

CShuwen committed:
level 8 - change Dates and times

CShuwen committed:
level 8 - change Dates and times

CShuwen committed:
merge branch 7 to master branch

CShuwen committed:
correct error: not able to detect empty deadline/event description

CShuwen committed:
Improve OOP for DUKE

CShuwen committed:
Refactor into Duke Package

CShuwen committed:
Merge branch 'add-gradle-support'

CShuwen committed:
Use Gradle

CShuwen committed:
JUnit Test implemented

CShuwen committed:
A-JavaDoc

CShuwen committed:
improve code based on coding standards

CShuwen committed:
improve code based on coding standards

CShuwen committed:
delete unnecessary file

shirsho-12 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,50 @@
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+
+public class Deadline extends Task{
```
Would recommend adding JavaDocs for this class.

shirsho-12 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,50 @@
+    }
+    @Override
+    public String toString(){
+        return "[D]["+ isDone + "]" + description + " (by " + this.by.format(DateTimeFormatter.ofPattern("MMM d yyyy")) + ")";
```
Maybe you could put MMM d yyyy as a constant.

shirsho-12 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,179 @@
+                                    "2001-02-10";
+                            throw new DukeException(err_msg);
+                        }
+                        continue;
+                    case "EVENT":
+                        String[] event_str_arr = input.split(" /");
+                        if (len <= 9 || event_str_arr.length <= 2 || event_str_arr[0].length() < 7) {
+                            String err_msg = "☹ OOPS!!! The description or date of a event cannot be empty";
+                            throw new DukeException(err_msg);
+                        }
+                        try {
+                            System.out.println(Arrays.toString(event_str_arr));
+                            LocalDate from = LocalDate.parse(event_str_arr[1]);
+                            LocalDate to = LocalDate.parse(event_str_arr[2]);
+                            if (from.isAfter(to)){
+                                String err_msg = "☹ OOPS!!! Your time range is from a date to another date that is earlier" +
+                                        "than the former. Please key in a valid time range";
+                                throw new DukeException(err_msg);
+                            }
+                            Event event = new Event(event_str_arr[0].substring(6), from, to, 0);
+                            ArrayList arraylist = storage.load();
+                            arraylist.add(event);
+                            System.out.println("Got it. I've added this task");
+                            storage.update_data(arraylist);
+
+                            System.out.println("added: " + event);
+                            System.out.println("Now you have " + arraylist.size() + " tasks in the list");
+                        } catch (DateTimeParseException e) {
+                            String err_msg = "☹ OOPS!!! The description or date for the event is wrong, plase key in the" +
+                                    "date in the format of yyyy-mm-dd, eg. 2001-02-10\n"
+                                    + "You may key in: event hw1 /2001-02-10 /2001-02-12, Duke.Duke will record your event hw1 as" +
+                                    "from 2001-02-10 to 2001-02-12";
+                            throw new DukeException(err_msg);
+                        }
+                        continue;
+                    case "ERROR":
+                        String err_msg = "☹ OOPS!!! I'm sorry, but I don't know what that means";
+                        throw new DukeException(err_msg);
+                    }
+                }catch (DukeException e){
```
Would strongly recommend splitting these into separate classes and packages for better extensibility. The current codebase looks very difficult to modify. Moreover, I'd suggest using enums.

shirsho-12 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,29 @@
+
+    public String parse(String input) {
+        Integer len = input.length();
+        if (input.equals("bye")){
+            return "BYE";
+        } else if (input.equals("list")) {
+            return "LIST";
+        } else if (len >= 8 && input.substring(0, 6).equals("delete")) {
+            return "DELETE";
+        } else if (len >= 6 && input.substring(0, 4).equals("mark")) {
+            return "MARK";
+        } else if (len >= 8 && input.substring(0, 6).equals("unmark")) {
+            return "UNMARK";
+        } else if (len >= 6 && input.substring(0, 4).equals("todo")) {
+            return "TODO";
+        } else if (len >= 10 && input.substring(0, 8).equals("deadline")) {
+            return "DEADLINE";
+        } else if (len >= 7 && input.substring(0, 5).equals("event")) {
+            return "EVENT";
+        } else {
+            return "ERROR";
+        }
```
Using enums here would simplify your logic greatly.

shirsho-12 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,78 @@
+    public ArrayList<Task> load(){
+        ArrayList<Task> arrayList = new ArrayList<>();
+        try{
+            File file = new File(data_address);
+            Scanner reader = new Scanner(file);
+            while (reader.hasNextLine()){
+                String data = reader.nextLine();
+                String[] data_parts = data.split("/");
+                String type = data_parts[0];
+                switch (type){
+                    case "T":
+                        arrayList.add(new ToDos(data_parts[2], Integer.valueOf(data_parts[1])));
+                        break;
+                    case "D":
+                        arrayList.add(new Deadline(data_parts[2], LocalDate.parse(data_parts[3]), Integer.valueOf(data_parts[1])));
+                        break;
+                    case "E":
+                        arrayList.add(new Event(data_parts[2], LocalDate.parse(data_parts[3]),LocalDate.parse(data_parts[4]), Integer.valueOf(data_parts[1])));
+                        break;
+                }
+            }
+            reader.close();
+        } catch (IOException e){
+            System.out.println(e);
+        }
+        return arrayList;
+    }
+
+    public void update_data(ArrayList<Task> arrayList){
+        String data = "";
+        try {
+            FileWriter fw = new FileWriter(this.data_address);
+            for (Task t: arrayList){
+                data += t.dataFormat() + "\n";
+            }
+            fw.write(data);
+            fw.close();
+        } catch (IOException E){
+            System.out.println(E);
+        }
```
I'd suggest adding JavaDocs to the 2 functions.

shirsho-12 reviewed (commented):
I generally like your code here, but would recommend further encapsulation and the use of enums to make your code more extensible.

CShuwen committed:
add "find" command to Duke

CShuwen committed:
merge A-Coding Standard into master branch

CShuwen committed:
Merge branch 'branch-Level-9'

CShuwen committed:
correct error in "FIND" command

shazxrin commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,195 @@
+                            System.out.println("Got it. I've added this task");
+                            storage.update_data(task_list);
+
+                            System.out.println("added: " + event);
+                            System.out.println("Now you have " + task_list.size() + " tasks in the list");
+                        } catch (DateTimeParseException e) {
+                            String err_msg = "☹ OOPS!!! The description or date for the event is wrong, plase key in the" +
+                                    "date in the format of yyyy-mm-dd, eg. 2001-02-10\n"
+                                    + "You may key in: event hw1 /2001-02-10 /2001-02-12, Duke.Duke will record your event hw1 as" +
+                                    "from 2001-02-10 to 2001-02-12";
+                            throw new DukeException(err_msg);
+                        }
+                        continue;
+                    case "FIND":
+                        String[] find_input_arr = input.split(" ");
+                        if (task_list.size() == 0){
+                            String err_msg = "You have not upload any task yet";
+                            throw new DukeException(err_msg);
+                        }
+                        System.out.println("Here are the matching tasks in your list:");
+                        i = 1;
+                        for (Task task : task_list) {
+                            if (task.description.contains(find_input_arr[1])){
+                                System.out.println(i + ". " + task);
+                                i++;
+                            }
+                        }
+                        if(i == 1) {
+                            System.out.println("There are no matching task");
+                        }
+                        continue;
+                    case "ERROR":
+                        String err_msg = "☹ OOPS!!! I'm sorry, but I don't know what that means";
+                        throw new DukeException(err_msg);
+                    }
+                }catch (DukeException e){
+                    System.out.println(e);
+            }
+            input = ui.readCommand();
+        }
```
According to the code quality standards, you need to [avoid long methods](https://nus-cs2103-ay2223s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-long-methods). I suggest splitting the handling of commands into a separate function or class.

shazxrin commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,91 @@
+ */
+public class Storage {
+    private final String data_address;
+    public Storage(String file_name) {
```
According to the coding standard about [naming variables](https://se-education.org/guides/conventions/java/intermediate.html#naming), variable names should be camelCase instead of snake_case.

shazxrin commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,71 @@
+ */
+public class Task {
+    protected String description;
+    private int isDone;
```
I suggest using `boolean` for isDone. The variable name may confuse other people thinking that it is a `boolean` based on the naming conventions from the coding standard.

shazxrin commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,91 @@
+     * @param arrayList an arraylist of tasks that need to be stored in duke.txt
+     * @throws IOException if the file duke.txt is not found.
+     */
+    public void update_data(ArrayList<Task> arrayList){
```
Similar to my comment on the variable names, function/method names should be camelCase instead of snake_case.

shazxrin commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,43 @@
+     * @return Type of the input/command in String form.
+     */
+    public String parse(String input) {
+        int len = input.length();
+        if (input.equals("bye")){
+            return "BYE";
+        } else if (input.equals("list")) {
+            return "LIST";
+        } else if (len >= 8 && input.substring(0, 6).equals("delete")) {
+            return "DELETE";
+        } else if (len >= 6 && input.substring(0, 4).equals("mark")) {
+            return "MARK";
+        } else if (len >= 8 && input.substring(0, 6).equals("unmark")) {
+            return "UNMARK";
+        } else if (len >= 6 && input.substring(0, 4).equals("todo")) {
+            return "TODO";
+        } else if (len >= 10 && input.substring(0, 8).equals("deadline")) {
+            return "DEADLINE";
+        } else if (len >= 7 && input.substring(0, 5).equals("event")) {
+            return "EVENT";
+        } else if (len >= 6 && input.substring(0, 4).equals("find")){
+            return "FIND";
+        } else {
+            return "ERROR";
+        }
```
The checking of length for a specific command in the input involves 'magic' numbers which the code quality document advised against. I suggest that you could check the command by splitting by the first whitespace encountered or split the input by the whitespace characters. For example:

```java
// input: "todo my description"
// inputTokens: ["todo", "my", "description"]
// command: "todo"
String[] inputTokens = input.split(" ");
String command = inputTokens[0];

if (command.equals("todo") {
    // ...
}
// ...
```

shazxrin reviewed (commented):
Some refactoring is required to comply to the coding and quality standards. It should be fine once they are rectified. 😃

tingyic commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,195 @@
+ * @version 1.0
+ * @since 0.0
+ */
+public class Duke {
```
Shouldn't all public classes have javadocs? 🤔

tingyic commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @Override
+    public void unMark(){
```
nitpicking but please add a space between reserved words and the Egyptian braces. I noticed the same minor issue in several other places also. 😓

tingyic commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,195 @@
+            ArrayList<Task> task_list = storage.load();
+            try {
+                switch (command){
+                    case "BYE" :
```
Shouldn't there be no indentation for case clauses?

tingyic commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,43 @@
+     * @param input input from user.
+     * @return Type of the input/command in String form.
+     */
+    public String parse(String input) {
```
That's a neat way to parse different commands! 👍

CShuwen committed:
Level-10 implement GUI

CShuwen committed:
add checkstyle and correct checkstyle errors

CShuwen committed:
Duke Class and Storage class: add Assertions

Some assumptions are made when the program is executed, checks by assertions are needed so that the runtime can verify them.

Let's check the size of the arraylist at several parts when new tasks are added.

When a new task has been successfully created, the array of task should contain more than one task.

CShuwen committed:
duke package: separate the package into two packages

Previous duke package contains too many classes that serve different purposes

Add an additional package called task and place Task class and its child classes into the package.

Improve clarity in purpose of each package.

CShuwen committed:
duke package: separate the package into two packages

Previous duke package contains too many classes that serve different purposes

Add an additional package called task and place Task class and its child classes into the package.

Improve clarity in purpose of each package.

CShuwen committed:
distribute classes to packages according to their usage

CShuwen committed:
improve codeQuality by removing over lengthy method

CShuwen committed:
Merge pull request #2 from CShuwen/branch-A-Assertions

Duke Class and Storage class: add Assertions

CShuwen committed:
merge master to branch-A-CodeQuality

CShuwen committed:
Merge pull request #3 from CShuwen/branch-A-CodeQuality

Branch a code quality

CShuwen committed:
add C-Undo extension

CShuwen committed:
add README file

CShuwen committed:
Add new theme to GUI, add personality to the bot

CShuwen committed:
delete necessary line

CShuwen committed:
add stylesheet for fxml file

CShuwen committed:
resolve duke response content error

CShuwen committed:
add Ui image

CShuwen committed:
delete unused file

CShuwen committed:
reduce profile image size

CShuwen committed:
delete unnecessary file

CShuwen committed:
correct checkstyle errors

CShuwen committed:
remove unnecessary build gradle setting

CShuwen committed:
update GUI font to accommodate mac users
